pkg/controllers/dashboardapi/settings/settings.go

```diff
@@ -148,5 +153,67 @@ func (s *settingsProvider) SetAll(settingsMap map[string]settings.Setting) error
 
 	s.fallback = fallback
 
+	if err := s.cleanupUnknownSettings(settingsMap); err != nil {
```

Comment:
I'm not sure if I agree. This was a deliberate decision to encapsulate and hide this in the concrete settings provider.
I don't see just yet what would be benefits of enlarging the surface of the provider interface to include this behavior (presumably by adding `CleanUknown` method).

Even from the naming standpoint `SetAll` (**set** is the operative word) can be read as make sure the settings store has **exactly** the settings that we defined and nothing else. In other words it can be seen as enforcing what was missing in the original implementation.

Reply:
I see Michael's point, but I think I prefer having the cleanupUnknownSettings call in here rather than exposing it and forcing something else to call that in addition to SetAll.  I don't think we'd ever want to SetAll without cleaning things up, right?

Reply:
> I don't see just yet what would be benefits of enlarging the surface of the provider interface to include this behavior (presumably by adding CleanUknown method).

The benefit is explicit control by the caller and clear semantics. The current code forces the caller to accept removed settings when using this method.

> Even from the naming standpoint SetAll (set is the operative word) can be read as make sure the settings store has exactly the settings that we defined and nothing else. In other words it can be seen as enforcing what was missing in the original implementation.

I don't agree on this part. At best, I can say that I think it's ambiguous. I definitely don't think that we can make a statement on the original intention of this method. The current comment also seems to be directly against this - it notes that it sets everything in the map, but makes no statement on settings not in the map.

> I don't think we'd ever want to SetAll without cleaning things up, right?

I think there's a world where the caller wants to set multiple settings that they control without function calls and so uses a method like this. That being said, no one uses this function for that purpose today so I don't think that we need to code for it.

To summarize my thoughts on this, I don't agree with the reasoning for this part of the change. That being said, I think we can go with this as long as we update the comment on both the struct implementation and on the interface to indicate that all settings not specified in the map will be dropped.

Reply:
> To summarize my thoughts on this, I don't agree with the reasoning for this part of the change.

It seems like a matter of difference of opinion. I'm not very strongly attached to the proposed solution but think that it addresses the original issue without "spilling" the semantics. I'm not convinced that we have enough evidence to support the idea that the caller will ever care about being able to choose whether to clean or not the unknown settings; thus making it premature engraving it in the interface IMHO.

That being said I 💯 agree and added comments describing the behavior of the provider implementation as well as stating in the interface that the behavior in regards to unknown settings is undefined and is up to a concrete implementation.